### PR TITLE
[2/N] Streaming Generator.  Support core worker APIs + cython generator interface.

### PR DESCRIPTION
**Pull request by rkooo567**
This reverts commit 122b70574cef011174262edce12c4b8444f65a53.

## Why are these changes needed?

This is the second PR to support streaming generator. 
- The detailed design and API proposal can be found from https://docs.google.com/document/d/1hAASLe2sCoay23raqxqwJdSDiJWNMcNhlTwWJXsJOU4/edit#heading=h.w91y1fgnpu0m. 
- The Execution plan can be found from https://docs.google.com/document/d/1hAASLe2sCoay23raqxqwJdSDiJWNMcNhlTwWJXsJOU4/edit#heading=h.kxktymq5ihf7. 

There will be 4 PRs to enable streaming generator for Ray Serve (phase 1). 
- [x] This PR -> introduce cpp interfaces to handle intermediate task return https://github.com/ray-project/ray/pull/35291
- [x] Support core worker APIs + cython generator interface. https://github.com/ray-project/ray/pull/35324 < --- this PR
- [ ] E2e integration https://github.com/ray-project/ray/pull/35325#pullrequestreview-1425581711
- [ ] Support async actors

This PR implements the Cython generator interface that users can use to obtain a next available object reference. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -198,6 +199,145 @@ class ObjectRefGenerator:
...
+                        if self._generator_task_completed_time is None:
+                            self._generator_task_completed_time = time.time()
+
+            # Currently, since the ordering of intermediate result report
+            # is not guaranteed, it is possible that althoug the task
+            # has succeeded, all of the object references are not reported
+            # (e.g., when there are network failures).
+            # If all the object refs are not reported to the generator
+            # within 30 seconds, we consider is as an unreconverable error.
+            if self._generator_task_completed_time:
```
Comment (rkooo567) [inline python/ray/_raylet.pyx:299]: Once we migrate to pubsub, this won't be necessary

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -198,6 +199,145 @@ class ObjectRefGenerator:
...
+            raise StopIteration
+
+    def __del__(self):
+        if hasattr(self.worker, "core_worker"):
+            # NOTE: This can be called multiple times
+            # because python doesn't guarantee __del__ is called
+            # only once.
+            self.worker.core_worker.delete_object_ref_stream(self._generator_ref)
+
+    def __getstate__(self):
```
Comment (rkooo567) [inline python/ray/_raylet.pyx:335]: this is to disallow passing generator to other tasks as an input or output

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -198,6 +199,145 @@ class ObjectRefGenerator:
...
+        raise TypeError(
+            "Serialization of the StreamingObjectRefGenerator "
```
Comment (rkooo567) [inline python/ray/_raylet.pyx:337]: TODO: improve the error message

File: python/ray/_raylet.pyx [hunk 4]
```diff
@@ -198,6 +199,145 @@ class ObjectRefGenerator:
...
+                    # It means the next wasn't reported although the task
+                    # has been terminated 30 seconds ago.
+                    self._generator_task_exception = AssertionError
+                    assert False, "Unexpected network failure occured."
+
+            if timeout_s != -1 and time.time() - last_time > timeout_s:
+                return ObjectRef.nil()
+
+            # 100us busy waiting
+            time.sleep(sleep_interval_s)
```
Comment (scv119) [inline python/ray/_raylet.pyx:311]: hmm can we do better than busy waiting?

Comment (rkooo567) [reply python/ray/_raylet.pyx:311]: I think it is possible, but I prefer to do it in a follow up PR. The current impl seems to work fine for serve so far, and it is not trivial to make it work with asyncio, so I need a bit of time to think about it.

File: python/ray/_raylet.pyx [hunk 5]
```diff
@@ -198,6 +199,145 @@ class ObjectRefGenerator:
...
+        if hasattr(self.worker, "core_worker"):
+            # NOTE: This can be called multiple times
+            # because python doesn't guarantee __del__ is called
+            # only once.
+            self.worker.core_worker.delete_object_ref_stream(self._generator_ref)
+
+    def __getstate__(self):
+        raise TypeError(
+            "Serialization of the StreamingObjectRefGenerator "
+            "is now allowed")
```
Comment (jjyao) [inline python/ray/_raylet.pyx:338]:
```suggestion
            "is not allowed")
```

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (jjyao) [reply python/ray/_raylet.pyx:299]: yea, it'd be nice to assume that the order is guaranteed.

**Review by jjyao — APPROVED** [review]

**Comment by ollie-iterators** [comment]
Lint error: 
!!! No API stability annotation found for:
--
  | ray.exceptions.ObjectRefStreamEoFError

File: python/ray/_raylet.pyx [hunk 6]
```diff
@@ -198,6 +198,149 @@ class ObjectRefGenerator:
...
+        try:
+            if hasattr(self.worker, "core_worker"):
+                obj = self.worker.core_worker.try_read_next_object_ref_stream(
+                    self._generator_ref)
+                return obj
+            else:
+                raise ValueError(
+                    "Cannot access the core worker. "
+                    "Did you already shutdown Ray via ray.shutdown()?")
+        except ObjectRefStreamEoFError:
```
Comment (edoakes) [inline python/ray/_raylet.pyx:328]: why not just use `StopIteration` directly?

File: python/ray/_raylet.pyx [hunk 7]
```diff
@@ -198,6 +198,149 @@ class ObjectRefGenerator:
...
+            raise StopIteration
```
Comment (edoakes) [inline python/ray/_raylet.pyx:329]:
```suggestion
            raise StopIteration from None
```

File: python/ray/_raylet.pyx (hunk 6 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:328]: cuz when there's an async actor, you need to raise StopAsyncIteration!

File: python/ray/_raylet.pyx (hunk 7 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:329]: is it because of stacktace?

**Pull request merged**